### PR TITLE
[MAINTENANCE] `execution_engine`  typing

### PR DESCRIPTION
**Pull request by Kilo59**
## Changes proposed in this pull request:
- Type-checking for the remaining `execution_engine` modules
- Note: the `spark` dependency is not yet part of the type-checking CI. Type-checking of spark related code is of limited value until this has been added.

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added [unit tests](https://docs.greatexpectations.io/docs/contributing/contributing_test#writing-unit-and-integration-tests) where applicable and made sure that new and existing tests are passing.
- [x] I have run any local integration tests and made sure that nothing is broken.

### CODE REVIEW

File: great_expectations/execution_engine/split_and_sample/data_sampler.py [hunk 1]
```diff
@@ -75,8 +79,8 @@ def verify_batch_spec_sampling_kwargs_key_exists(
     def get_sampling_kwargs_value_or_default(
         batch_spec: BatchSpec,
         sampling_kwargs_key: str,
-        default_value: Optional[Any] = None,
-    ) -> Optional[Any]:
+        default_value: Optional[T] = None,
+    ) -> T | Any:
```
Comment (NathanFarmer) [inline great_expectations/execution_engine/split_and_sample/data_sampler.py:83]: Should this be returning `T` or `Any`? Does `T` not one of `Any`?

Comment (NathanFarmer) [inline great_expectations/execution_engine/split_and_sample/data_sampler.py:83]: Should this be returning `T` or `Any`? Is `T` not a member of `Any`?

File: great_expectations/execution_engine/split_and_sample/sparkdf_data_splitter.py [hunk 2]
```diff
@@ -141,7 +143,7 @@ def split_on_date_parts(
         """
         self._validate_date_parts(date_parts)
 
-        date_parts: List[DatePart] = self._convert_date_parts(date_parts)
+        date_parts = self._convert_date_parts(date_parts)
```
Comment (NathanFarmer) [inline great_expectations/execution_engine/split_and_sample/sparkdf_data_splitter.py:146]: In this case I would prefer to rename the variable and keep the accurate type hint.

File: great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py [hunk 3]
```diff
@@ -173,7 +173,7 @@ def split_on_date_parts(
         """
         self._validate_date_parts(date_parts)
 
-        date_parts: List[DatePart] = self._convert_date_parts(date_parts)
+        date_parts = self._convert_date_parts(date_parts)
```
Comment (NathanFarmer) [inline great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py:179]: Same comment...we have more information that we could convey after calling this method (`date_parts` can no longer be `List[str]`).

File: great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py [hunk 4]
```diff
@@ -366,13 +366,13 @@ def get_data_for_batch_identifiers(
...
             )
         else:
-            batch_identifiers_list: List[
-                dict
-            ] = self.get_data_for_batch_identifiers_for_non_date_part_splitters(
-                execution_engine,
-                table_name,
-                processed_splitter_method_name,
-                splitter_kwargs,
+            batch_identifiers_list = (
```
Comment (NathanFarmer) [inline great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py:373]: Shouldn't we declare `batch_identifiers_list: List[dict]` above the start of the `if`/`else` block?

File: great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py [hunk 5]
```diff
@@ -490,7 +490,7 @@ def get_split_query_for_data_for_batch_identifiers_for_split_on_date_parts(
         """
         self._validate_date_parts(date_parts)
 
-        date_parts: List[DatePart] = self._convert_date_parts(date_parts)
+        date_parts = self._convert_date_parts(date_parts)
```
Comment (NathanFarmer) [inline great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py:497]: Same comment regarding variable rename.

File: great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py [hunk 6]
```diff
@@ -526,19 +526,18 @@ def get_split_query_for_data_for_batch_identifiers_for_split_on_date_parts(
                         )
                     )
 
-                concat_clause: List[Label] = [
+                concat_clause = [
```
Comment (NathanFarmer) [inline great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py:535]: Shouldn't we declare this type above the `if`/`else` block? If i'm reading this code I'm going to skip the `if` block, get to this variable, then need to scan above and look in the unrelated code path to find the type.

File: great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py [hunk 7]
```diff
@@ -526,19 +526,18 @@ def get_split_query_for_data_for_batch_identifiers_for_split_on_date_parts(
...
                     sa.func.distinct(concat_date_parts).label("concat_distinct_values"),
                 ]
             else:
-                concat_date_parts: concat = sa.func.concat(
+                concat_date_parts = sa.func.concat(
```
Comment (NathanFarmer) [inline great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py:539]: It would be helpful to have this type declared above the `if`/`else` block in my opinion.

File: great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py [hunk 8]
```diff
@@ -548,7 +547,7 @@ def get_split_query_for_data_for_batch_identifiers_for_split_on_date_parts(
                         ),
                     )
 
-                concat_clause: List[Label] = [
+                concat_clause = [
```
Comment (NathanFarmer) [inline great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py:556]: Same comment

File: great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py [hunk 9]
```diff
@@ -631,7 +633,7 @@ def _get_params_for_batch_identifiers_from_date_part_splitter(
         Returns:
             List of dicts of the form [{column_name: {date_part_name: date_part_value}}]
         """
-        date_parts: List[DatePart] = self._convert_date_parts(date_parts)
+        date_parts = self._convert_date_parts(date_parts)
```
Comment (NathanFarmer) [inline great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py:642]: Same comment

**Review by NathanFarmer — COMMENTED** [review]
A few questions from me. I only didn't approve because auto-merge was turned on.

File: great_expectations/execution_engine/split_and_sample/data_sampler.py (hunk 1 above)
Comment (Kilo59) [reply great_expectations/execution_engine/split_and_sample/data_sampler.py:83]: `T` is only representing the `default_value`, if it is provided.
`Any` represents the type of value returned if the "sampling kwargs" are "found".

File: great_expectations/execution_engine/split_and_sample/sparkdf_data_splitter.py (hunk 2 above)
Comment (Kilo59) [reply great_expectations/execution_engine/split_and_sample/sparkdf_data_splitter.py:146]: The type  of `date_parts` is already declared in the method signature.
```python
date_parts: Union[List[DatePart], List[str]],
```
 It is "narrowed" to `List[DatePart]` at line 146 but we can't re-declare types like that.

I don't have a problem with the change you describe but I really try to avoid renaming or making runtime changes as part of these initial type-checkings PRs.
It makes it much clearer when going through the blame that no runtime changes were made.

File: great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py (hunk 4 above)
Comment (Kilo59) [reply great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py:373]: I'm happy to do that if that's a pattern we prefer.
I try to make as little change as possible in these initial type coverage PRs so I generally only make changes to make the type-checker happy and avoid stylistic changes.

File: great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py (hunk 6 above)
Comment (Kilo59) [reply great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py:535]: That makes sense to me.

File: great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py (hunk 9 above)
Comment (Kilo59) [reply great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py:642]: I'd rather not create new variables as part of this PR.

File: great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py (hunk 3 above)
Comment (Kilo59) [reply great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py:179]: https://github.com/great-expectations/great_expectations/pull/6730/files#r1064677778

File: great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py (hunk 5 above)
Comment (Kilo59) [reply great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py:497]: https://github.com/great-expectations/great_expectations/pull/6730/files#r1064677778

File: great_expectations/execution_engine/split_and_sample/sparkdf_data_splitter.py (hunk 2 above)
Comment (Kilo59) [reply great_expectations/execution_engine/split_and_sample/sparkdf_data_splitter.py:146]: ☝️ If you still feel strongly that you'd like to see a new variable I will make the change.

File: great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py (hunk 4 above)
Comment (Kilo59) [reply great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py:373]: I've pushed these changes here and elsewhere.

File: great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py (hunk 6 above)
Comment (Kilo59) [reply great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py:535]: Done

File: great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py (hunk 7 above)
Comment (Kilo59) [reply great_expectations/execution_engine/split_and_sample/sqlalchemy_data_splitter.py:539]: Done

**Review by NathanFarmer — APPROVED** [review]
LGTM!
